components/tooltip/script.js

```diff
@@ -40,30 +42,70 @@ class Tooltip extends HTMLElement {
         return overflows;
     }
 
+    get async() {
+        return this.hasAttribute('async');
+    }
+
     /**
-     * Sets or changes the tooltip message.
-     * Resolve the Promise and if it is a regular function, just set the returned value.
-     * Directly set the message if the provided argument is not a function.
-     * @param {Promise|string|number} message
-     */
-    set message(message) {
-        if (typeof message === 'function') {
-            this._messageSlot.textContent = 'Loading...';
+     * Sets the textContent of the message slot HTMLElement
+     * @param {string | number} content
+    */
+    setTooltipContent(content) {
+        return this._messageSlot.textContent = content;
+    }
 
-            const messageResult = message();
+    /**
+     * Checks the validity of a synchronous message.
+     * Allowed types are - string | number | Function<string | number>.
+     * @param {any} value
+     * @returns {boolean} - true if it is valid and false if not.
+    */
+    isSyncMessageValid(value) {
+        const allowedSyncTypes = ['string', 'number', 'function'];
+        const type = typeof value;
 
-            if (messageResult instanceof Promise) {
-                messageResult.then(result => {
-                    this._messageSlot.textContent = result;
-                });
-            } else { // Handle when a regular function is passed.
-                this._messageSlot.textContent = messageResult;
-            }
+        if (!allowedSyncTypes.includes(type)) return false;
+        if (type === 'function' && (value() instanceof Promise)) return console.error('Using Promise in sync mode is forbidden. Add "async" attribute to the gameface-tooltip element.');
+        return true;
+    }
 
-            return;
-        }
+    /**
+     * Synchronously sets the value of the tooltip.
+     * @param {string | number} value
+    */
+    setSyncMessage(value) {
+        if (!this.isSyncMessageValid(value)) return;
 
-        this._messageSlot.textContent = message;
+        value = (typeof value === 'function') ? value() : value;
+        this.setTooltipContent(value);
+    }
+
+    /**
+     * Asynchronously sets the value of the tooltip.
+     * @param {Function<Promise>} value
+     * @returns {Promise} - a promise that will resolve with the value that was set.
+    */
+    setAsyncMessage(value) {
+        if (typeof value !== 'function') return console.error(NOT_A_PROMISE_ERROR);
+        const valueResult = value();
+        if (!(valueResult instanceof Promise)) return console.error(NOT_A_PROMISE_ERROR)
+
+        this.setTooltipContent('Loading...');

```

Comment:
Maybe we can let the user define the loading content instead just displaying 'Loading...'. We can add an optional parameter to the setAsyncMessage(value, loadingMessage = 'Loading...') and for the setMessage too.

But what if the user call the setMessage(callback, '')? then the tooltip maybe should not be visible? Is this is for another task or we can do it in this?

Reply:
Sounds like another task to me.